File: 530. Minimum Absolute Difference in BST/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        self.res = 100001
        self.prev = -1
        self.dfs(root)
        return self.res

    def dfs(self, node):
        if node.left: self.dfs(node.left)
        if self.prev > -1:
            self.res = min(self.res, node.val - self.prev)
        self.prev = node.val
        if node.right: self.dfs(node.right)

File: 530. Minimum Absolute Difference in BST/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 100001, prev = -1;
    int getMinimumDifference(TreeNode* root) {
        dfs(root);
        return res;
    }
    void dfs(TreeNode* node) {
        if (node->left != nullptr) dfs(node->left);
        if (prev > -1) res = min(res, node->val - prev);
        prev = node->val;
        if (node->right != nullptr) dfs(node->right);
    }
};
